include/rapidcheck/state/Commands.hpp

```diff
@@ -1,9 +1,109 @@
 #pragma once
 
 #include "rapidcheck/detail/Results.h"
+#include "rapidcheck/state/gen/ParallelCommands.hpp"
+#include "rapidcheck/Assertions.h"
+#include <thread>
+#include <mutex>
+#include "rapidcheck/detail/Barrier.h"
 
 namespace rc {
 namespace state {
+namespace detail {
+
+
+template <typename Model, typename Cmd>
+struct CommandResult
+{
+  CommandResult(const std::shared_ptr<const Cmd>& command,
+                std::function<void(const Model&)>&& verifyFunc)
+    : command(command)
+    , verifyFunc(std::move(verifyFunc))
+  {
+  }
+
+  std::shared_ptr<const Cmd> command;
+  std::function<void(const Model&)> verifyFunc;
+};
+
+template <typename Model, typename Cmd>
+struct ParallelExecutionResult {
+  std::vector<CommandResult<Model, Cmd>> prefix;
+  std::vector<CommandResult<Model, Cmd>> left;
+  std::vector<CommandResult<Model, Cmd>> right;
+};
+
+template <typename Model, typename Cmd>
+bool hasValidInterleaving(const std::vector<CommandResult<Model, Cmd>> &left,
+                          const std::vector<CommandResult<Model, Cmd>> &right,
+                          std::size_t leftIndex,
+                          std::size_t rightIndex,
+                          const Model &state) {
+  const auto hasLeft = leftIndex < left.size();
+  const auto hasRight = rightIndex < right.size();
+
+  // Base case, no more commands to verify
+  if (!hasRight && !hasLeft) {
+    return true;
+  }
+
+  if (hasLeft) {
+    try {
+      auto preState = state;
+      auto currentState = state;
+      auto commandResult = left[leftIndex];
+      commandResult.command->apply(currentState);
+      commandResult.verifyFunc(preState);
+      if (hasValidInterleaving(
+              left, right, leftIndex + 1, rightIndex, currentState)) {
+        return true;
+      }
+    } catch (...) {
+      // Failed
+    }
+  }
+
+  if (hasRight) {
+    try {
+      auto preState = state;
+      auto currentState = state;
+      auto commandResult = right[rightIndex];
+      commandResult.command->apply(currentState);
+      commandResult.verifyFunc(preState);
+      if (hasValidInterleaving(
+              left, right, leftIndex, rightIndex + 1, currentState)) {
+        return true;
+      }
+    } catch (...) {
+      // Failed
+    }
+  }
+
+  return false;
+}
+
+template <typename Model, typename Cmd>
+void verifyExecution(
+    const ParallelExecutionResult<Model, Cmd> &executionResult,
+    const Model &state) {
+  auto currentState = state;
+  // Verify prefix.
+  for (const auto &commandResult : executionResult.prefix) {
+    auto preState = currentState;
+    commandResult.command->apply(currentState);
+    commandResult.verifyFunc(preState);
+  }
+
+  // verify parallel sequences
+  if (!hasValidInterleaving(
+        executionResult.left, executionResult.right, 0, 0, currentState)) {
+    RC_FAIL("No possible interleaving");
```

Comment:
Agree but I'm not sure what to say here (been trying to find something better). Not that it matters but QuickCheck for Erlang outputs "Reason: no_possible_interleaving" (https://youtu.be/zi0rHwfiX1Q?t=35m3s)